### DISCUSSION TITLE
DRY implementation

### DISCUSSION THREAD

**Discussion by momshaddinury** [EXERLOG/exer_log#114] [Ideas]
### [DRY development](https://www.digitalocean.com/community/tutorials/what-is-dry-development)

In current implementation things like Theme provider, Network status checker are breaking DRY. Looking for suggestions to implement them using BaseView. The challenge here is that we need access to riverpod as well. 

1. While creating BaseView class, keep ConsumerWIdget and ConsumerStatefulWidget in consideration. (Hard path)
2. Creating a normal BaseView (StatelessWidget) and only using the Consumer wrapper for Riverpod. (Easy)

**Comment by jorre127** [comment]
@dinurymomshad I'm not that familiar with riverpods, mostly use Provider. So I don't know if this would be possible in riverpods, but in most of my project we use a custom widget called ProviderWidget, this provides all the things necessary for that page like theme and viewmodel (which is the controller in our case). That widget would look something like this:

```
ProviderWidget<AuthenticationController>(
builder: (context, theme, controller) => // contents of the page
) 
```

Are you thinking about something similar? Or do you mean something else with a BaseView.

**Reply by momshaddinury** [reply]
@jorre127 https://github.com/dinurymomshad/listify/blob/main/lib/core/base/base_view.dart
Something like this but lot simpler

**Reply by jorre127** [reply]
@dinurymomshad Okay I see what you mean now. That looks like a pretty good implementation. Only, because it changes the way you make widgets (using overriding of different methods to create you screen) it might make it harder for newer people to jump in and make changes. So I propose to take a more standard approach and put all that logic of the baseView class you showed and just put it in a widget. we'd have to only put that BaseView widget once on every screen so it might go against DRY a little, but I think it's worth it for the sake of clarity, what do you think?

**Reply by momshaddinury** [reply]
I agree with you 100%. Just wondering if we could just use riverpod's provider to provide these things where ever we need them.

**Reply by jorre127** [reply]
@dinurymomshad Something like this? Or do you mean something else.

```
BaseView<AuthenticationController>(
create: () => ref.read(AuthenticationController.controller)
builder: (context, controller) => //widget can now user controller provided by riverpods.
)
```

**Reply by momshaddinury** [reply]
I will share something today. Gotta check your theme implementation

**Reply by momshaddinury** [reply]
**KISS** should also be considered when implementing **DRY**. I doing some homework on these things. Let's see what I can come up with.
